### PR TITLE
feat(device_selection): Implement popup

### PR DESCRIPTION
**Pull request by hristoterezov**

### CODE REVIEW

File: react/features/base/dialog/components/Dialog.web.js [hunk 1]
```diff
@@ -31,6 +27,11 @@ class Dialog extends AbstractDialog {
         isModal: React.PropTypes.bool,
 
         /**
+         * Disables rendering of the submit button.
+         */
+        submitDisabled: React.PropTypes.bool,
+
```
Comment (damencho) [inline react/features/base/dialog/components/Dialog.web.js]: We were using wrong property and I have fixed this in another PR (https://github.com/jitsi/jitsi-meet/pull/1632). We have a property okDisabled defined in the Abstract dialog and it was only speaker stats that was using submitDisabled, so I got rid of it. Maybe we can merge this PR and then I fill fix the merge conflict in my PR with fixing this.

File: react/features/base/dialog/components/NonReduxDialog.web.js [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+         * Width of the dialog, can be:
+         * - 'small' (400px), 'medium' (600px), 'large' (800px),
+         * 'x-large' (968px)
+         * - integer value for pixel width
+         * - string value for percentage
+         */
+        width: React.PropTypes.string
+
+    };
+
```
Comment (damencho) [inline react/features/base/dialog/components/NonReduxDialog.web.js]: So basically you have duplicated the properties from the Abstract dialog? Can we discuss and find some approach and have them in one place so we do not duplicate them.

File: react/features/base/dialog/components/NonReduxDialog.web.js [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        /**
+         * The handler for the event when submitting the dialog.
+         */
+        onSubmit: React.PropTypes.func,
+
+        /**
+         * Disables rendering of the submit button.
+         */
+        submitDisabled: React.PropTypes.bool,
+
```
Comment (damencho) [inline react/features/base/dialog/components/NonReduxDialog.web.js]: See my comment above, we do not need submitDisabled.

File: react/features/base/dialog/components/NonReduxDialog.web.js (hunk 2 above)
Comment (hristoterezov) [inline react/features/base/dialog/components/NonReduxDialog.web.js]: I can define them somewhere as a const and then use them for both components? WDYT?

File: react/features/base/dialog/components/Dialog.web.js (hunk 1 above)
Comment (hristoterezov) [inline react/features/base/dialog/components/Dialog.web.js]: Sounds good!

File: react/features/base/dialog/components/NonReduxDialog.web.js (hunk 2 above)
Comment (damencho) [inline react/features/base/dialog/components/NonReduxDialog.web.js]: Yep, I was talking for this with @lyubomir and he purposed the same, like having constants.js ...

File: react/features/base/dialog/components/Dialog.web.js (hunk 1 above)
Comment (damencho) [inline react/features/base/dialog/components/Dialog.web.js]: Yep, I'm wrong. But anyway I think we can rename submitDisabled, cause we have okDisabled and this is confusing ... but this is not to be fixed in this PR.

File: react/features/base/dialog/components/NonReduxDialog.web.js (hunk 2 above)
Comment (hristoterezov) [inline react/features/base/dialog/components/NonReduxDialog.web.js]: Fixed!

File: react/features/toolbox/defaultToolbarButtons.js [hunk 4]
```diff
@@ -153,12 +154,32 @@ export default {
...
+
+            return openDialOutDialog();
         },
         tooltipKey: 'dialOut.dialOut'
     },
 
     /**
+     * The descriptor of the device selection toolbar button.
+     */
+    deviceselection: {
```
Comment (yanas) [inline react/features/toolbox/defaultToolbarButtons.js]: Hm. I think that we need to use the current Settings button here. It already has some logic that make it open the device selection dialog if (interfaceConfig.SETTINGS_SECTIONS.length === 1 && UIUtil.isSettingEnabled('devices'))

File: react/features/base/dialog/components/NonReduxDialog.web.js [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+import React, { Component } from 'react';
+
+import { translate } from '../../i18n';
+
+import { dialogPropTypes } from '../constants';
+
+/**
+ * Web dialog that uses atlaskit modal-dialog to display dialogs.
+ */
+class NonReduxDialog extends Component {
```
Comment (yanas) [inline react/features/base/dialog/components/NonReduxDialog.web.js]: I would suggest to name this StatelessDialog.

File: react/features/device-selection/actions.js [hunk 6]
```diff
@@ -1,18 +1,173 @@
...
+ * @returns {{
+ *     type: SET_DEVICE_SELECTION_POPUP,
+ *     popupDialog: Object
+ * }}
+ */
+function _setDeviceSelectionPopup(popupDialog) {
+    return {
+        type: SET_DEVICE_SELECTION_POPUP,
+        popupDialog
+    };
```
Comment (yanas) [inline react/features/device-selection/actions.js]: As we discussed off-github can we rename this to popupDialogData or something similar that would give us an indication that this isn't a component, but some backend instead.

File: react/features/toolbox/defaultToolbarButtons.js (hunk 4 above)
Comment (hristoterezov) [inline react/features/toolbox/defaultToolbarButtons.js]: We agreed offline that it's better to have this new button because the whole logic is pretty tight to the device selection only for now.

File: react/features/device-selection/actions.js [hunk 7]
```diff
@@ -1,18 +1,173 @@
...
+ *
+ * @param {Dispatch} dispatch - The redux {@code dispatch} function.
+ * @param {Function} getState - The redux function that gets/retrieves the redux
+ * state.
+ * @param {Object} request - The request to be processed.
+ * @param {Function} responseCallback - The callback that will send the
+ * response.
+ * @returns {boolean}
+ */ // eslint-disable-next-line max-params
+function _processRequest(dispatch, getState, request, responseCallback) {
```
Comment (yanas) [inline react/features/device-selection/actions.js]: Sort functions alphabetically.

File: react/features/device-selection/popup.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import 'aui-css';
+import 'aui-experimental-css';
```
Comment (yanas) [inline react/features/device-selection/popup.js]: Do we use those 2 anywhere?

Comment (hristoterezov) [inline react/features/device-selection/popup.js]: Yes. For the fonts, etc. Without them the dialog looks different.

Comment (yanas) [inline react/features/device-selection/popup.js]: Just checked and we can easily remove them for meet in general. However this will require some more testing and we'll definitely do this in a separate PR.

File: react/features/toolbox/defaultToolbarButtons.js (hunk 4 above)
Comment (yanas) [inline react/features/toolbox/defaultToolbarButtons.js]: Let's just rename it to something that suggests it's filmstrip-only button. Something like: fo-deviceselection. We can find a better way to do this in the future.
